### DISCUSSION TITLE
Using Matrix Profiles for Forecasting and Deep Learning

### DISCUSSION THREAD

**Discussion by seanlaw** [TDAmeritrade/stumpy#164] [Q&A]
@sn88798 wrote:

> Hi @seanlaw ,
Thanks for your nice work on stumpy.
I am a machine learning practitioner with keen interest on time series
I am keen on using stumpy to train my models & then make inference on unseen data

> Regarding time series prediction, could you kindly provide some example notebooks on how to use time series chains for prediction ? (currently I am using state space methods for forecasting)
Regarding regime change, could you kindly provide example notebooks on how to use FLUSS for this purpose ? (currently I am using Hidden Markov model + Deep Learning methods)
thanks !

**Comment by seanlaw** [comment]
@sn88798 Thank you for the kind words and for using STUMPY! I moved your questions over here (from #155 ) so that we can have a longer discussion here without distracting from the core point of the other issue. I certainly don't have all of the answers but would you mind telling providing more information on the following:

1. What use case are you forecasting/predicting (i.e., can you share with me what your time series data actually contains)?
2. Have you already taken a look at the [time series chains notebook](https://stumpy.readthedocs.io/en/latest/Tutorial_Time_Series_Chains.html)? Have you also tried using [FBProphet](https://facebook.github.io/prophet/) for forecasting?
3. Have you already taken a look at the [FLUSS/FLOSS notebook](https://stumpy.readthedocs.io/en/latest/Tutorial_Semantic_Segmentation.html)? I don't have any experience using HMM or DL for examining regime changes but perhaps we can focus on what problem you are trying to solve first and then figure out an approach that might help (i.e., FLUSS/FLOSS may not be suitable depending on your use case).

**Comment by sn88798** [comment]
Hi @seanlaw,
thanks for quick response
I am a regular Kaggle competitor, so I would quote you some use cases
1) time series prediction --> the often quoted Rossmann sales forecast & also recent M5 Walmart Sales forecast (competition is still on). Given each store location, for each product ID predict sales for next 28 days. Training data is usually about 1 to 2 years of daily sales data. Meta-data likes sales item, whether day is holiday or not is also indicated.

3) For classification (recent competition "ion channel"), time series look like chart below:
Top panel is Current Amplitude. Bottom panel is # of open channels (based on current amplitude)
So given this training data, have to predict for test data the # of open channels (given only current amplitude chart)

![time series example](https://user-images.githubusercontent.com/23212334/80314395-fc848e80-8823-11ea-827a-24cc1081a7d2.png)

I have briefly looked through the stumpy tutorials. generating the matrix profile should not be an issue. But I am not clear how to use the matrix profile as features for my model (eg random forest or deep neural network).  
The traditional way to feature engineer a time series is to extract statistics from it. Example rolling mean, standard deviation, whether day is normal or special (ie holiday) etc...

thanks for help

**Comment by seanlaw** [comment]
I am no expert here but the first things that I would try is consider adding the matrix profile and, maybe more importantly, the matrix profile indices (including the left and right indices) into your feature engineering. So, for each subsequence, you'd could add the following types of features (that may or may not be useful):

1. the z-normalized Euclidean distance to your global nearest neighbor (which tells you how similar this subsequence is to the nearest neighbor) 
2. the global location of your nearest neighbor to your left (this is the left matrix profile index)
3. the global location of your nearest neighbor to your right (this is the right matrix profile index)
4. the relative location of your nearest neighbor to your left (this is the left matrix profile index minus the index of current subsequence)
5. the relative location of your nearest neighbor to your right (this is the right matrix profile index minus the index of the current subsequence)

I think the last two are most interesting as they provide context of where its nearest neighbor is relative to one self (i.e., is my nearest neighbor 100 time units away on my left?). You might be able to do some sort of binning (of the relative location of the indices) and maybe apply a one-hot encoding where it makes sense.  

Again, just throwing out some ideas here. Perhaps, these [two](https://www.cs.ucr.edu/~eamonn/WeaklyLabeledTimeSeries.pdf) [papers](https://www.cs.ucr.edu/~eamonn/Efficient%20and%20Effective%20Labeling%20of%20Massive%20Entomological%20Datasets.pdf) regarding matrix profiles may be relevant (I haven't read them yet).

**Comment by seanlaw** [comment]
Also, say, for the top 10-20 motifs (i.e., a subsequence with one of the top 10-20 lowest matrix profile values), go through each subsequence in your time series and establish a one-hot encoding for how far (z-norm Euclidean distance) this subsequence is to each of the 10-20 motifs.  

Additionally, you can compare subsequences before a motif as well.

I haven't put too much thought into this but, in the "current amplitude" case, the concept of a [multidimensional matrix profile](https://www.cs.ucr.edu/~eamonn/Motif_Discovery_ICDM.pdf) but your mileage may vary. might be useful (see `mstump`/`mstumped`). 

Are you able to share your Kaggle handle/profile with me? I'd love to check out your competitions!

**Comment by seanlaw** [comment]
@sn88798 Closing this for now. Feel free to re-open if you have more questions or comments. Thank you for using STUMPY!
